cmd/opampsupervisor/supervisor/packages.go

```diff
@@ -0,0 +1,412 @@
+	}
+
+	err = agentFile.Truncate(0)
+	if err != nil {
+		return fmt.Errorf("truncate agent file: %w", err)
+	}
+
+	// Write new agent to existing agent file.
+	// We only copy up to maxAgentBytes to avoid compression bombs.
+	_, err = io.CopyN(agentFile, tar, maxAgentBytes)
+	switch {
+	case errors.Is(err, io.EOF): // OK
+	case err != nil:
+		restoreErr := renameFile(agentBackupPath, p.agentExePath)
+		return errors.Join(fmt.Errorf("write package to file: %w", err), restoreErr)
+	default:
+		restoreErr := renameFile(agentBackupPath, p.agentExePath)
+		return errors.Join(fmt.Errorf("agent package met or exceeded %d bytes", maxAgentBytes), restoreErr)
+	}
+
+	return nil
+}
+
+func (p *packageManager) DeletePackage(packageName string) error {
+	if packageName != agentPackageKey {
+		// We only take the agent package, so the package already doesn't exist.
+		return nil
+	}
+
+	// We will never delete the agent package.
+	return errors.New("cannot delete top-level package")
+}
+
+func (p *packageManager) LastReportedStatuses() (*protobufs.PackageStatuses, error) {
+	lastStatusBytes, err := os.ReadFile(p.lastPackageStatusPath())
+	switch {
+	case errors.Is(err, os.ErrNotExist):
+		// No package statuses exists
+		return nil, nil
+	case err != nil:
+		return nil, fmt.Errorf("read last package statuses: %w", err)
+	}
+
+	var ret protobufs.PackageStatuses
+	err = proto.Unmarshal(lastStatusBytes, &ret)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal last package statuses: %w", err)
+	}
+
+	return &ret, nil
+}
+
+func (p packageManager) SetLastReportedStatuses(statuses *protobufs.PackageStatuses) error {
+	lastStatusBytes, err := proto.Marshal(statuses)
+	if err != nil {
+		return fmt.Errorf("marshal statuses: %w", err)
+	}
+
+	err = os.WriteFile(p.lastPackageStatusPath(), lastStatusBytes, 0o600)
+	if err != nil {
+		return fmt.Errorf("write package statues: %w", err)
+	}
+
+	return nil
+}
+
+func (p *packageManager) lastPackageStatusPath() string {
+	return filepath.Join(p.storageDir, lastPackageStatusFileName)
+}
+
+func verifyPackageHash(packageBytes, expectedHash []byte) error {
+	actualHash := sha256.Sum256(packageBytes)
+	if !bytes.Equal(actualHash[:], expectedHash) {
+		return errors.New("invalid hash for package")
+	}
+
+	return nil
+}
+
+func parsePackageSignature(signature []byte) (b64Cert, b64Signature []byte, err error) {
+	splitSignature := bytes.SplitN(signature, []byte(" "), 2)
+	if len(splitSignature) != 2 {
+		return nil, nil, fmt.Errorf("signature must be formatted as a space separated cert and signature")
+	}
+
+	return splitSignature[0], splitSignature[1], nil
+}
+
+// verifyPackageSignature verifies that the b64Signature is a valid signature for packageBytes.
+// b64Cert is used to validate the identity of the signature against the identities in the
+// provided checkOpts.
+func verifyPackageSignature(ctx context.Context, checkOpts *cosign.CheckOpts, packageBytes, b64Cert, b64Signature []byte) error {
```

Comment:
This is the function that detects if a malicious OpAMP servers has sent us a bad executable, right? Can we document this in a bit more detail, particularly how VerifyBlobSignature works.